File: frontend/src/types/session.ts
export type InitialChoice = "KNOW" | "NOT_KNOW" | "HAS_QUESTION"
export type VoiceInputTarget = "SELF_EXPLANATION" | "GUIDED_ANSWER" | "DOUBT" | "APPEAL"

export type SessionStatus = "IN_PROGRESS" | "COMPLETED" | "STOPPED_LIMIT" | "NEED_HUMAN" | "PAUSED"

export type FlowStage =
  | "WAIT_INITIAL_CHOICE"
  | "CAPTURING_INPUT"
  | "TRANSCRIBING"
  | "AI_EVALUATING"
  | "WAIT_STUDENT_ACTION"
  | "WAIT_GUIDED_ANSWERS"
  | "SHOWING_FULL_SOLUTION"

export interface ErrorEvidence {
  quote: string
  locationDescription: string
  reason: string
  thinkingDirection: string
}

export interface AIEvaluation {
  id: number
  correctness: "CORRECT" | "WRONG" | "UNCERTAIN"
  completeness: "COMPLETE" | "INCOMPLETE"
  coveredPoints: number[]
  errorEvidence: ErrorEvidence[]
  needHumanReason: string | null
  promptVersion: string
  modelProvider: string
  modelName: string
  createdAt: string
}

export interface SupportEvent {
  id: number
  supportType: "ASK_FOCUSED_QUESTION" | "GIVE_HINT" | "GIVE_CORRECTION" | "CORRECT_AND_ASK"
  round: number
  status: "VALID" | "REFUSED"
  content: string
  supportKind: "EVALUATION" | "GUIDED_QUESTIONS" | "SIMPLE_DOUBT" | "CURRENT_STEP"
  mainDraft: string | null
  doubtText: string | null
  guidedQuestions: GuidedQuestion[] | null
  guidedAnswers: GuidedAnswer[] | null
  followUpContent: string | null
  createdAt: string
}

export interface GuidedQuestion {
  id: string
  question: string
}

export interface GuidedAnswer {
  questionId: string
  answer: string
}

export interface LearningTimelineItem {
  id: string
  eventType: "SUBMISSION" | "EVALUATION" | "SUPPORT" | "FULL_SOLUTION" | "NEED_HUMAN"
  speaker: "STUDENT" | "AI" | "SYSTEM"
  submissionType: "SELF_EXPLANATION" | "SUPPORT_REQUEST" | "GUIDED_ANSWER" | "DOUBT" | "APPEAL" | null
  content: string
  correctness: AIEvaluation["correctness"] | null
  completeness: AIEvaluation["completeness"] | null
  action: "COMPLETE" | "NEED_HUMAN" | SupportEvent["supportType"] | null
  createdAt: string
}

export interface Session {
  id: number
  questionId: number
  status: SessionStatus
  flowStage: FlowStage
  round: number
  supportCountRound: number
  supportCountTotal: number
  noProgressCount: number
  noProgressHelpRequestCount: number
  solutionExposed: boolean
  completionType: "INDEPENDENT" | "WITH_SUPPORT" | "AFTER_SOLUTION" | null
  coveredPointsCurrentRound: string[]
  coveredPointsAll: string[]
  currentDraft: string
  version: number
  initialChoice: InitialChoice | null
  needHumanReason: string | null
  latestEvaluation: AIEvaluation | null
  latestSupport: SupportEvent | null
  teachingGeneration:
    | { status: "NOT_REQUIRED" }
    | { status: "SUCCEEDED"; supportEventId: number }
    | null
}

File: backend/app/schemas/session.py
from datetime import datetime
from typing import Literal

from pydantic import Field

from app.schemas.ai_evaluation import AIEvaluationResponse, Completeness, Correctness
from app.schemas.question import QuestionSchema, RequiredText
from app.schemas.support import GuidedAnswer, SupportEventResponse

SessionStatus = Literal["IN_PROGRESS", "COMPLETED", "STOPPED_LIMIT", "NEED_HUMAN", "PAUSED"]
FlowStage = Literal[
    "WAIT_INITIAL_CHOICE",
    "CAPTURING_INPUT",
    "TRANSCRIBING",
    "AI_EVALUATING",
    "WAIT_STUDENT_ACTION",
    "WAIT_GUIDED_ANSWERS",
    "SHOWING_FULL_SOLUTION",
]
InitialChoice = Literal["KNOW", "NOT_KNOW", "HAS_QUESTION"]
VoiceInputTarget = Literal["SELF_EXPLANATION", "GUIDED_ANSWER", "DOUBT", "APPEAL"]


class CreateSessionInput(QuestionSchema):
    question_id: int = Field(gt=0)


class InitialChoiceInput(QuestionSchema):
    choice: InitialChoice
    version: int = Field(ge=0)


class TextAttemptInput(QuestionSchema):
    confirmed_text: RequiredText = Field(max_length=4000)
    version: int = Field(ge=0)
    voice_attempt_id: int | None = Field(default=None, gt=0)


class StudentActionInput(QuestionSchema):
    version: int = Field(ge=0)


class HelpRequestInput(StudentActionInput):
    main_draft: str = ""


class DoubtRequestInput(HelpRequestInput):
    doubt_text: RequiredText
    voice_attempt_id: int | None = Field(default=None, gt=0)


class GuidedAnswersInput(StudentActionInput):
    answers: list[GuidedAnswer] = Field(min_length=1)
    voice_attempt_id: int | None = Field(default=None, gt=0)


class AppealInput(StudentActionInput):
    reason: RequiredText
    voice_attempt_id: int | None = Field(default=None, gt=0)


class SolutionUnderstandingInput(StudentActionInput):
    understood: bool


TimelineEventType = Literal[
    "SUBMISSION",
    "EVALUATION",
    "SUPPORT",
    "FULL_SOLUTION",
    "NEED_HUMAN",
]
TimelineSpeaker = Literal["STUDENT", "AI", "SYSTEM"]
TimelineSubmissionType = Literal[
    "SELF_EXPLANATION",
    "SUPPORT_REQUEST",
    "GUIDED_ANSWER",
    "DOUBT",
    "APPEAL",
]
TimelineSupportType = Literal[
    "ASK_FOCUSED_QUESTION",
    "GIVE_HINT",
    "GIVE_CORRECTION",
    "CORRECT_AND_ASK",
]
TimelineAction = Literal["COMPLETE", "NEED_HUMAN"] | TimelineSupportType


class LearningTimelineItemResponse(QuestionSchema):
    id: str
    event_type: TimelineEventType
    speaker: TimelineSpeaker
    submission_type: TimelineSubmissionType | None
    content: str
    correctness: Correctness | None
    completeness: Completeness | None
    action: TimelineAction | None
    created_at: datetime


class SessionResponse(QuestionSchema):
    id: int
    question_id: int
    initial_choice: InitialChoice | None
    status: SessionStatus
    flow_stage: FlowStage
    round: int
    support_count_round: int
    support_count_total: int
    no_progress_count: int
    no_progress_help_request_count: int
    solution_exposed: bool
    completion_type: str | None
    need_human_reason: str | None
    covered_points_current_round: list[str]
    covered_points_all: list[str]
    current_draft: str
    paused_from_stage: FlowStage | None
    version: int
    created_at: datetime
    updated_at: datetime
    finished_at: datetime | None
    latest_evaluation: AIEvaluationResponse | None = None
    latest_support: SupportEventResponse | None = None
